website/blog/deployers.md

```diff
@@ -0,0 +1,610 @@
+
+1. **Components**. A deployer starts weavelets and tells them which components
+   to host. A deployer also ensures that weavelets know the addresses of other
+   weavelets. If a deployer starts a new weavelet, for example, the deployer
+   notifies all other weavelets of its existence, including its address and the
+   components its hosting. Conversely, if a deployer detects that a weavelet has
+   failed, the deployer notifies all other weavelets of its failure.
+
+2. **Listeners**. When a component calls the [Instance.Listener][Listener]
+   method to request a network listener, a deployer picks an address for the
+   listener and ensures that the listener is publicly accessible. Multiple
+   weavelets may share the same listener, and a deployer must ensure that
+   traffic is balanced across them. This often involves running or configuring a
+   proxy.
+
+3. **Telemetry**. A deployer collects, aggregates, and exports all telemetry
+   produced by weavelets. This includes logs, metrics, traces, and profiles.
+
+A deployer and a weavelet communicate by exchanging protocol buffers other over
+a pair of Unix pipes.  We call the part of a deployer that communicates with a
+weavelet an **envelope**. In the appendix of this article, we describe the
+low-level details of the envelope-weavelet protocol, but for most deployers, it
+should suffice to use ServiceWeaver's [Envelope][] API.
+
+Communication between an envelope and a weavelet is either weavelet initiated or
+envelope initiated. Weavelet initiated communication shows up as a method call
+to an [EnvelopeHandler][] interface supplied by the deployer implementation.
+
+```go
+type EnvelopeHandler interface {
+    // Responsibility 1: Components.
+    ActivateComponent(context.Context, *protos.ActivateComponentRequest) (*protos.ActivateComponentReply, error)
+
+    // Responsibility 2: Listeners.
+    GetListenerAddress(context.Context, *protos.GetListenerAddressRequest) (*protos.GetListenerAddressReply, error)
+    ExportListener(context.Context, *protos.ExportListenerRequest) (*protos.ExportListenerReply, error)
+
+    // Responsibility 3: Telemetry.
+    HandleLogEntry(context.Context, *protos.LogEntry) error
+    HandleTraceSpans(context.Context, []trace.ReadOnlySpan) error
+}
+```
+
+Envelope initiated communication is performed by invoking a method on
+`Envelope`.
+
+```go
+// Responsibility 1: Components.
+func (e *Envelope) UpdateRoutingInfo(routing *protos.RoutingInfo) error
+func (e *Envelope) UpdateComponents(components []string) error
+
+// Responsibility 3: Telemetry.
+func (e *Envelope) GetHealth() protos.HealthStatus
+func (e *Envelope) GetMetrics() ([]*metrics.MetricSnapshot, error)
+func (e *Envelope) GetLoad() (*protos.LoadReport, error)
+func (e *Envelope) GetProfile(req *protos.GetProfileRequest) ([]byte, error)
+```
+
+## A Simple Multiprocess Deployer
+
+In this section, we implement a fully working multiprocess deployer. To make
```

Comment:
Maybe make "a fully working multiprocess deployer" the link and drop the "See here for the full source code" sentence.